homeassistant/components/keyboard_remote/__init__.py

```diff
@@ -48,165 +60,210 @@
+                self.handlers_by_descriptor[descriptor] = handler
+            else:
+                name = dev_block.get(DEVICE_NAME)
+                self.handlers_by_name[name] = handler
+
+    async def setup(self):
+        """Start inotify watching for events, start event monitoring for those already connected, and start monitoring for device connection/disconnection."""
+
+        # start watching
+        self.watcher = aionotify.Watcher()
+        self.watcher.watch(
+            alias="devinput",
+            path=DEVINPUT,
+            flags=aionotify.Flags.CREATE
+            | aionotify.Flags.ATTRIB
+            | aionotify.Flags.DELETE,
+        )
+        await self.watcher.setup(self.hass.loop)
 
-class KeyboardRemoteThread(threading.Thread):
-    """This interfaces with the inputdevice using evdev."""
+        # add initial devices (do this AFTER starting watcher in order to avoid race conditions leading to missing device connections)
+        descriptors = list_devices(DEVINPUT)
+        for descriptor in descriptors:
+            dev, handler = self.get_device_handler(descriptor)
 
-    def __init__(self, hass, device_name, device_descriptor, key_value):
-        """Construct a thread listening for events on one device."""
-        self.hass = hass
-        self.device_name = device_name
-        self.device_descriptor = device_descriptor
-        self.key_value = key_value
-
-        if self.device_descriptor:
-            self.device_id = self.device_descriptor
-        else:
-            self.device_id = self.device_name
-
-        self.dev = self._get_keyboard_device()
-        if self.dev is not None:
-            _LOGGER.debug("Keyboard connected, %s", self.device_id)
-        else:
-            _LOGGER.debug(
-                "Keyboard not connected, %s. " "Check /dev/input/event* permissions",
-                self.device_id,
-            )
-
-            id_folder = "/dev/input/by-id/"
-
-            if os.path.isdir(id_folder):
-                from evdev import InputDevice, list_devices
-
-                device_names = [
-                    InputDevice(file_name).name for file_name in list_devices()
-                ]
-                _LOGGER.debug(
-                    "Possible device names are: %s. "
-                    "Possible device descriptors are %s: %s",
-                    device_names,
-                    id_folder,
-                    os.listdir(id_folder),
-                )
+            if handler is None:
+                continue
 
-        threading.Thread.__init__(self)
-        self.stopped = threading.Event()
-        self.hass = hass
+            self.active_handlers_by_descriptor[descriptor] = handler
+            await handler.async_start_monitoring(dev)
 
-    def _get_keyboard_device(self):
-        """Get the keyboard device."""
-        from evdev import InputDevice, list_devices
+        self.hass.loop.create_task(self.async_monitor_devices())
```

Comment:
We should stop monitoring when home assistant stops. Register a listener for home assistant stop event.